operator/es_client.go

```diff
@@ -350,29 +350,37 @@ func (esSettings *ESSettings) updateRebalance(value string) {
 // repeatedly query shard allocations to ensure success of drain operation.
 func (c *ESClient) waitForEmptyEsNode(ctx context.Context, pod *v1.Pod) error {
 	podIP := pod.Status.PodIP
+
+	// Counter to track the number of retries
+	retryCount := 0
+
 	_, err := resty.New().
 		SetRetryCount(c.DrainingConfig.MaxRetries).
 		SetRetryWaitTime(c.DrainingConfig.MinimumWaitTime).
 		SetRetryMaxWaitTime(c.DrainingConfig.MaximumWaitTime).
 		AddRetryCondition(
-			// It is expected to return (bool, error) pair. Resty will retry
-			// in case condition returns true or non nil error.
-			func(r *resty.Response) (bool, error) {
+			// It is expected to return bool. Resty will retry in case condition returns true.
+			func(r *resty.Response, err error) bool {
+				retryCount++
+				log.Debugf("Waiting for Elasticsearch node to remove all shards. Details: Namespace=%s, PodName=%s, PodIP=%s, RetryCount=%d.",
+					pod.Namespace, pod.Name, podIP, retryCount)
 				select {
 				case <-ctx.Done():
-					// Return false to not retry and return the context error directly.
-					return false, ctx.Err()
+					// Return false to not retry
+					return false
```

Comment:
Is it not possible to forward errors anymore in the new version? Will it cause a logical issue if we start considering this as successful now?

Reply:
Imagine that we retry for 999 times and we fail all the time. We will log 998 times in the retry condition function and restry will return the last error (999th) and the caller will log it.

Reply:
My question is about the value returned by waitForEmptyEsNode() . In this scenario it seems the value would not be an error since the error is not in our control. In the previous version, we used to send an error mentioning the context was cancelled. Based on which the callers and rest of es operator code was supposed to simply exit the loop. Will this behaviour be the same or change?

Reply:
It's the same.

Reply:
What will be the output of the function `waitForEmptyEsNode()` ? Is it going to return an `err` or `nil`? If it is returning `err` then it is fine. We would simply exit the operator loop. But if not that can lead to many problems.

For now we can just add a unit test for this case (send a context and cancel the Drain()) and the expected output it. But ideally an e2e/integration test is required here.

For e.g. if we are returning `nil` without waiting for shards to actually relocate - We may just prematurely delete the pod and lose some shards. - https://github.com/zalando-incubator/es-operator/blob/master/operator/operator.go#L373-L386 Let's not end up in a weird state.